A79-non-per-call-metrics-architecture.md

```diff
@@ -0,0 +1,418 @@
+   template <typename T>
+   void HistogramRecord(GlobalDoubleHistogramHandle handle, T value,
+                        RefCountedPtr<LabelValueSet> label_values,
+                        RefCountedPtr<LabelValueSet> optional_values);
+
+   // Use the stats plugin to get a representation of label values that can be
+   // saved for multiple uses later.
+   RefCountedPtr<LabelValueSet> MakeLabelValueSet(
+       absl::Span<absl::string_view> label_values);
+ };
+
+ void RegisterStatsPlugin(std::shared_ptr<StatsPlugin> plugin);
+
+ // Invoked to get a stats plugin list for a specified scope.
+ static StatsPluginGroup GetStatsPluginsForChannel(ChannelScope scope);
+
+private:
+ std::vector<std::shared_ptr<StatsPlugin>> plugins_;
+};
+```
+
+##### Java
+
+(To be filled)
+
+##### Go
+
+(To be filled)
+
+### Metric Instrument Naming Conventions
+
+gRPC metrics follow the same general
+[guidelines](https://opentelemetry.io/docs/specs/semconv/general/metrics/#general-guidelines)
+as OpenTelemetry. The following additional guidelines apply -
+
+*   Metrics defined by gRPC should go under the `grpc.` namespace.
+*   Third-party plugins to gRPC should NOT use the `grpc.` namespace.
+*   Newer metrics should follow the same style as existing metrics. * Metrics
+    from load-balancing policies should be nested under `grpc.lb.`, e.g.
+    `grpc.lb.pick_first.`.
+*   Metrics from resolvers should be nested under `grpc.resolver.`, e.g.
+    `grpc.resolver.dns.`.
+*   Metrics from transports should be nested under `grpc.transport.`, e.g.,
+    `grpc.transport.http2.`
+
+### Metric Stability Guidelines and Guarantees
+
+From a stability perspective, metrics exported by gRPC will be in one of the
+following categories -
+
+*   Experimental, off by default
+*   Stable, off by default
+*   Stable, on by default
+
+Experimental metrics should always be off-by-default. gRPC reserves the right to
+change or remove these metrics. Documentation for the given metric determines
+whether the metric is experimental or non-experimental. An alternative approach
+of using an `experimental` namespace was considered, but not preferred since
+de-experimentalization of the metric would result in disruption of the
+corresponding metrics data (timeseries).
+
+New metrics should start as experimental. Once they are stable, they can
+optionally be made on-by-default. The distinction between on-by-default and
+off-by-default metrics allows gRPC to support metrics that may be desirable in
+some cases but too expensive to collect by default.
+
+### Changes to gRPC OpenTelemetry Plugin
+
+Optional Labels - Users should be able to configure gRPC OpenTelemetry plugins
+to record the additional optional labels. One suggested approach of implementing
+this is for the plugin instance to store a bit set for each instrument where
+each bit denotes whether the plugin should record the label in the corresponding
+position of the set of optional labels.
```

Comment:
A few questions:

1. Are "the additional optional labels" be configured here the optional label keys registered with the `GlobalInstrumentsRegistry` API or it's the `optional_label_key` added through the `OpenTelemetryPluginBuilder` API?
2. "to store a bit set for each instrument", it sounds like this is to configure the optional label keys registered with the `GlobalInstrumentsRegistry` API since each instrument has different set of optional label keys?
3. If 2 is the case, then what would this API look like? I'm guessing the user of the `OpenTelemetryPluginBuilder` API needs to set this bitset for the instrument that they are interested in configuring, but would they have access to the `GlobalInstrumentHandle` at the point of configuration or they can use the `GlobalInstrumentsRegistry::ForEach` API and use the index to map to the bitset (I guess that probably would work)?